promql/ast.go

```diff
@@ -0,0 +1,344 @@
+}
+
+// RecordStmt represents an added recording rule.
+type RecordStmt struct {
+	Name   string
+	Expr   Expr
+	Labels clientmodel.LabelSet
+}
+
+func (*AlertStmt) stmt()  {}
+func (*EvalStmt) stmt()   {}
+func (*RecordStmt) stmt() {}
+
+// ExprType is the type an evaluated expression returns.
+type ExprType int
+
+const (
+	ExprNone ExprType = iota
+	ExprScalar
+	ExprVector
+	ExprMatrix
+	ExprString
+)
+
+func (e ExprType) String() string {
+	switch e {
+	case ExprNone:
+		return "<ExprNone>"
+	case ExprScalar:
+		return "scalar"
+	case ExprVector:
+		return "vector"
+	case ExprMatrix:
+		return "matrix"
+	case ExprString:
+		return "string"
+	}
+	panic("promql.ExprType.String: unhandled expression type")
+}
+
+// Expr is a generic interface for all expression types.
+type Expr interface {
+	Node
+
+	// Type returns the type the expression evaluates to. It does not perform
+	// in-depth checks as this is done at parsing-time.
+	Type() ExprType
+	// expr ensures that no other types accidentally implement the interface.
+	expr()
+}
+
+// Expressions is a list of expression nodes that implements Node.
+type Expressions []Expr
+
+// AggregateExpr represents an aggregation operation on a vector.
+type AggregateExpr struct {
+	Op              itemType               // The used aggregation operation.
+	Expr            Expr                   // The vector expression over which is aggregated.
+	Grouping        clientmodel.LabelNames // The labels by which to group the vector.
+	KeepExtraLabels bool                   // Whether to keep extra labels common among result elements.
+}
+
+// BinaryExpr represents a binary expression between two child expressions.
+type BinaryExpr struct {
+	Op       itemType // The operation of the expression.
+	LHS, RHS Expr     // The operands on the respective sides of the operator.
+
+	// The matching behavior for the operation if both operands are vectors.
+	// If they are not this field is nil.
+	VectorMatching *VectorMatching
+}
+
+// Call represents a function call.
+type Call struct {
+	Func *Function   // The function that was called.
+	Args Expressions // Arguments used in the call.
+}
+
+// MatrixSelector represents a matrix selection.
+type MatrixSelector struct {
+	Name          string
+	Interval      time.Duration
+	Offset        time.Duration
+	LabelMatchers metric.LabelMatchers
+
+	// The series iterators are populated at query analysis time.
+	iterators map[clientmodel.Fingerprint]local.SeriesIterator
+	metrics   map[clientmodel.Fingerprint]clientmodel.COWMetric
+	// Fingerprints are populated from label matchers at query analysis time.
+	fingerprints clientmodel.Fingerprints
+}
+
+// NumberLiteral represents a number.
+type NumberLiteral struct {
+	N clientmodel.SampleValue
+}
+
+// ParenExpr wraps an expression so it cannot be disassembled as a consequence
+// of operator precendence.
+type ParenExpr struct {
+	Expr Expr
+}
+
+// StringLiteral represents a string.
+type StringLiteral struct {
+	S string
```

Comment:
I'd still go for the full `String` name here. Or `Value`. Dito for the `NumberLiteral`. Increased clarity when referred to in other contexts.